pkg/ccl/changefeedccl/cdctest/validator.go

```diff
@@ -381,6 +353,175 @@ func (v *beforeAfterValidator) Failures() []string {
 	return v.failures
 }
 
+type mvccTimestampValidator struct {
+	failures []string
+}
+
+// NewMvccTimestampValidator returns a Validator that verifies that the emitted
+// row includes the mvcc_timestamp field and that its value is not later than
+// the updated timestamp
+func NewMvccTimestampValidator() Validator {
+	return &mvccTimestampValidator{}
+}
+
+// NoteRow implements the Validator interface.
+func (v *mvccTimestampValidator) NoteRow(
+	partition, key, value string, updated hlc.Timestamp, topic string,
+) error {
+	valueJSON, err := json.ParseJSON(value)
+	if err != nil {
+		return err
+	}
+
+	mvccJSON, err := valueJSON.FetchValKey("mvcc_timestamp")
+	if err != nil {
+		return err
+	}
+	if mvccJSON == nil {
+		v.failures = append(v.failures, fmt.Sprintf(
+			"expected MVCC timestamp, got nil"))
+		return nil
+	}
+
+	mvccJSONText, err := mvccJSON.AsText()
+	if err != nil {
+		return err
+	}
+
+	if *mvccJSONText > updated.AsOfSystemTime() {
+		v.failures = append(v.failures, fmt.Sprintf(
+			`expected MVCC timestamp to be earlier or equal to updated timestamp (%s), got %s`,
+			updated.AsOfSystemTime(), *mvccJSONText))
+	}
+
+	return nil
+}
+
+// NoteResolved implements the Validator interface.
+func (v *mvccTimestampValidator) NoteResolved(partition string, resolved hlc.Timestamp) error {
+	return nil
+}
+
+// Failures implements the Validator interface.
+func (v *mvccTimestampValidator) Failures() []string {
+	return v.failures
+}
+
+type keyInValueValidator struct {
+	primaryKeyCols []string
+	failures       []string
+}
+
+// NewKeyInValueValidator returns a Validator that verifies that the emitted row
+// includes the key inside a field named "key" inside the value. It should be
+// used only when key_in_value is specified in the changefeed.
+func NewKeyInValueValidator(sqlDB *gosql.DB, table string) (Validator, error) {
+	primaryKeyCols, err := fetchPrimaryKeyCols(sqlDB, table)
+	if err != nil {
+		return nil, errors.Wrap(err, "fetchPrimaryKeyCols failed")
+	}
+
+	return &keyInValueValidator{
+		primaryKeyCols: primaryKeyCols,
+	}, nil
+}
+
+// NoteRow implements the Validator interface.
+func (v *keyInValueValidator) NoteRow(
+	partition, key, value string, updated hlc.Timestamp, topic string,
+) error {
+	keyJSON, err := json.ParseJSON(key)
+	if err != nil {
+		return err
+	}
+	keyJSONAsArray, ok := keyJSON.AsArray()
+	if !ok || len(keyJSONAsArray) != len(v.primaryKeyCols) {
+		return errors.Errorf(
```

Comment:
should probably append to v.failures here instead of returning error. ditto for (some but not all) other errors in this method